Implement Datastore, PUT, and POST methods

gferioli0418 opened the pull request:
POM.XML: include dependency for datastore

Entity.java: include static methods to get all events or a single event. Also updated to get data from datastore. Included a static method to get a list of events with a certain cityId.

EntityServlet.Java: include datastore. Included POST method which also serves as a PUT method. 

City.java: include static methods to get all cities or a single city. Also updated to get data from datastore.

CityServlet.Java: include datastore. Included POST method which also serves as a PUT method. 

Country.java: include static methods to get all countries or a single country. Also updated to get data from datastore.

CountryServlet.Java: include datastore. Included POST method which also serves as a PUT method.

gferioli0418 committed:
Implement Datastore, PUT, and POST methods

josueetcom commented on src/main/java/com/google/sps/travelbud/City.java:
```diff
@@ -1,37 +1,96 @@
+          cityEvents.get(entityId));
+      cities.add(c);
+    }
+    return cities;
```
This is not an actual suggestion (this may actually be slower) but one way to use a `List<Event>` instead of a map could've been :
```java
    PreparedQuery queryEvents = datastore.prepare(new Query("Events"));
    List<Event> events = new ArrayList<>();

    // iterate through each event entity
    for (Entity entity : events.asIterable()) {
      long cityId = (long) entity.getProperty("cityId");
      Event e = new Event((long) entity.getKey().getId(), (String) entity.getProperty("name"),
          (long) entity.getProperty("cityId"), (String) entity.getProperty("description"),
          (String) entity.getProperty("date"), (String) entity.getProperty("location"),
          (double) entity.getProperty("pricing"));
      events.add(e);
    }

    List<City> cities = new ArrayList<>();
    Query queryCity = new Query("City");
    PreparedQuery citiesQuery = datastore.prepare(queryCity);
    // iterate throuch each city entity and add it to the list;
    for (Entity entity : citiesQuery.asIterable()) {
      long entityId = (long) entity.getKey().getId();
      City c = new City(entityId, (String) entity.getProperty("name"),
          (long) entity.getProperty("countryId"), (String) entity.getProperty("description"),
          events.stream().filter(e -> e.getCityId() == entityId).collect(Collectors.toList()));
      cities.add(c);
    }
    return cities;
```

josueetcom commented on src/main/java/com/google/sps/travelbud/City.java:
```diff
@@ -1,37 +1,96 @@
+    // iterate through each event entity
+    for (Entity entity : events.asIterable()) {
+      long cityId = (long) entity.getProperty("cityId");
+      Event e = new Event((long) entity.getKey().getId(), (String) entity.getProperty("name"),
+          (long) entity.getProperty("cityId"), (String) entity.getProperty("description"),
+          (String) entity.getProperty("date"), (String) entity.getProperty("location"),
+          (double) entity.getProperty("pricing"));
```
Suggestion: To make this easier to read and use, I suggest either adding a constructor for the `City`, `Country`, and `Event` classes that take in an `Entity` or providing static `fromEntity()` functions in those classes to do this for you. E.g. 

Option 1: Another constructor
```java
City(Entity entity) {
    this.id = entity.getKey().getId();
    this.name = (String) entity.getProperty("name");
    this.countryId = (long) entity.getProperty("countryId");
    this.description = (String) entity.getProperty("description");
}
```

Option 2: A helper static method
```java
static City fromEntity(Entity e) {
    return new City(entity.Key().getId(), (String) entity.getProperty("name"),
          (long) entity.getProperty("countryId"), (String) entity.getProperty("description"),
          new ArrayList<Event>())
}
```

That way the callers don't have to reason about passing parameters in the right order or what datastore key corresponds with what field property and lines like the one highlighted can be replaced with `Event e = new Event(entity);` or `Event e = Event.fromEntity(entity);`

gferioli0418 replied:
This is a great idea! thanks for the tip!

josueetcom commented on src/main/java/com/google/sps/travelbud/City.java:
```diff
@@ -1,37 +1,96 @@
+  }
+
+  public static City getCity(DatastoreService datastore, long id) {
+    // first search for events with cityID
+    List<Event> cityEvents = Event.getEventsInCity(datastore, id);
+
+    Query queryCity = new Query("City");
+    PreparedQuery cities = datastore.prepare(queryCity);
+    // iterate through each city event
+    for (Entity entity : cities.asIterable()) {
+      long entityId = (long) entity.getKey().getId();
+      // only return the city with the right id
+      if (entityId == id) {
+        City c = new City(entityId, (String) entity.getProperty("name"),
+            (long) entity.getProperty("countryId"), (String) entity.getProperty("description"),
+            cityEvents);
+        return c;
+      }
+    }
+    return null;
```
Per [documentation](https://cloud.google.com/appengine/docs/standard/java/datastore/creating-entities#Retrieving_entities) and docs for the [Key](https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/datastore/Key) class, it looks like this could be as simple as:

```java
    Key key = KeyFactory.createKey("City", id);
    try {
        Entity entity = datastore.get(key);
        return City.fromEntity(entity);
    } catch (EntityNotFoundException e) {
        return null;
    }
```

(assuming you go the `fromEntity()` static method option mentioned above)

josueetcom commented on src/main/java/com/google/sps/travelbud/CityServlet.java:
```diff
@@ -25,42 +31,54 @@
+    long id = Long.parseLong(getParameter(request, "id", ""));
+    String name = getParameter(request, "name", "");
+    long countryId = Long.parseLong(getParameter(request, "countryId", ""));
+    String description = getParameter(request, "description", "");
 
-      if (tempCityId == cityId) {
-        return tempCity;
-      }
+    Entity cityEntity = new Entity("City", id);
```
I suggest not reading an ID from the poster as there's no guarantee it'll be unique. Instead, you can rely on the ID generated when you do the put (see [`Entity(java.lang.String kind)`](https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/datastore/Entity.html#Entity-java.lang.String-))

josueetcom commented on src/main/java/com/google/sps/travelbud/CityServlet.java:
```diff
@@ -25,42 +31,54 @@
-        return tempCity;
-      }
+    Entity cityEntity = new Entity("City", id);
+    cityEntity.setProperty("name", name);
+    cityEntity.setProperty("countryId", countryId);
+    cityEntity.setProperty("description", description);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.put(cityEntity);
```
I have no objections here as it's pretty simple, but you _might_ (not required) consider refactoring this into a method like `City.createCity(String name, String countryId, String description)`

josueetcom commented on src/main/java/com/google/sps/travelbud/CountryServlet.java:
```diff
@@ -48,18 +57,31 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
+    // Get the input from the form.
+    long id = Long.parseLong(getParameter(request, "id", ""));
+    String name = getParameter(request, "name", "");
+    List<String> cultureDos = Arrays.asList(getParameter(request, "cultureDos", "").split(",", -1));
+    List<String> cultureDonts =
+        Arrays.asList(getParameter(request, "cultureDonts", "").split(",", -1));
```
Interesting. This means there can't be any commas in the list values. It looks like this though instead of splitting on `,` it might be better to have the front end figure out how to dynamically add more values to the same `cultureDos` and `cultureDonts` keys so that you can use [`HttpServletRequest.getParameterValues()`](https://docs.oracle.com/javaee/7/api/javax/servlet/ServletRequest.html#getParameterValues-java.lang.String-) instead:
```java
    List<String> cultureDos = Arrays.asList(request.getParameterValues("cultureDos"));
    List<String> cultureDonts = Arrays.asList(request.getParameterValues("cultureDonts"));
```

josueetcom commented on src/main/java/com/google/sps/travelbud/CountryServlet.java:
```diff
@@ -48,18 +57,31 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
-  public Country getCountry(List<Country> countries, String path) {
-    // adjust endpoint
-    String countryName = path.substring(1).toLowerCase();
+  @Override
```
Missing a blank line above

josueetcom commented on src/main/java/com/google/sps/travelbud/Event.java:
```diff
@@ -30,24 +26,85 @@
     this.pricing = pricing;
   }
-  int getId() {
+  public static List<Event> getEventsInCity(DatastoreService datastore, long cityId) {
```
Missing blank line above

josueetcom commented on src/main/java/com/google/sps/travelbud/Event.java:
```diff
@@ -30,24 +26,85 @@
+    }
+    return events;
+  }
+  public static List<Event> getAll(DatastoreService datastore) {
```
Missing blank line above

josueetcom reviewed (changes requested):
The comments left in `City.java` and `CityServlet.java` are mostly applicable in the other files as well.

gferioli0418 committed:
Made code more readable

gferioli0418 committed:
Pass checks

gferioli0418 committed:
Pass checks

gferioli0418 committed:
Pass checks

gferioli0418 committed:
Pass checks

josueetcom commented on src/main/java/com/google/sps/travelbud/Event.java:
```diff
@@ -26,6 +29,15 @@
     this.location = location;
     this.pricing = pricing;
   }
+  Event(Entity entity) {
+    this.id = (long) entity.getKey().getId();
+    this.name = (String) entity.getProperty("name");
+    this.cityId = (long) entity.getProperty("cityId");
+    this.description = (String) entity.getProperty("description");
+    this.date = (String) entity.getProperty("date");
+    this.location = (String) entity.getProperty("location");
+    this.pricing = (double) entity.getProperty("pricing");
+  }
```
Technically there should be a blank line above and beneath this constructor
```suggestion

  Event(Entity entity) {
    this.id = (long) entity.getKey().getId();
    this.name = (String) entity.getProperty("name");
    this.cityId = (long) entity.getProperty("cityId");
    this.description = (String) entity.getProperty("description");
    this.date = (String) entity.getProperty("date");
    this.location = (String) entity.getProperty("location");
    this.pricing = (double) entity.getProperty("pricing");
  }

```

josueetcom commented on src/main/java/com/google/sps/travelbud/EventServlet.java:
```diff
@@ -1,3 +1,5 @@
+
+
```
??

josueetcom reviewed (dismissed):
LGTM, though I think you want to merge this PR to the `master` branch?

gferioli0418 committed:
Pass checks

gferioli0418 committed:
Merge branch 'GET_request' into Datastore/PUT/POST

gferioli0418 committed:
Pass Checks

gferioli0418 committed:
Make local Datastor/PUT/POST branch up to date

Merge branch 'Datastore/PUT/POST' of https://github.com/gferioli0418/travelbud into Datastore/PUT/POST

gferioli0418 committed:
Pass Checks
